src/tree/context-dep.h

```diff
@@ -99,16 +100,57 @@ class ContextDependency: public ContextDependencyInterface {
   /// GetPdfInfo returns a vector indexed by pdf-id, saying for each pdf which
   /// pairs of (phone, pdf-class) it can correspond to.  (Usually just one).
   /// c.f. hmm/hmm-topology.h for meaning of pdf-class.
-  virtual void GetPdfInfo(const std::vector<int32> &phones,  // list of phones
-                  const std::vector<int32> &num_pdf_classes,  // indexed by phone,
-                  std::vector<std::vector<std::pair<int32, int32> > > *pdf_info)
+  /// This is the old, simpler interface of GetPdfInfo(), and that this one can
+  /// only be called if the HmmTopology object's IsHmm() function call returns
+  /// true.
+  virtual void GetPdfInfo(
+      const std::vector<int32> &phones,  // list of phones
+      const std::vector<int32> &num_pdf_classes,  // indexed by phone,
+      std::vector<std::vector<std::pair<int32, int32> > > *pdf_info)
+      const;
+
+  /// This function outputs information about what possible pdf-ids can
+  /// be generated for HMM-states; it covers the general case where
+  /// the self-loop pdf-class may be different from the forward-transition
+  /// pdf-class, so we are asking not about the set of possible pdf-ids
+  /// for a given (phone, pdf-class), but the set of possible ordered pairs
+  /// (forward-transition-pdf, self-loop-pdf) for a given (phone,
+  /// forward-transition-pdf-class, self-loop-pdf-class).
+  /// Note: 'phones' is a list of integer ids of phones, and
+  /// 'pdf-class-pairs', indexed by phone, is a list of pairs
+  /// (forward-transition-pdf-class, self-loop-pdf-class) that we can have for
+  /// that phone.
+  /// The output 'pdf_info' is indexed first by phone and then by the
+  /// same index that indexes each element of 'pdf_class_pairs',
+  /// and tells us for each pair in 'pdf_class_pairs', what is the
+  /// list of possible (forward-transition-pdf-id, self-loop-pdf-id) that
+  /// we can have.
+  /// This is less efficient than the other version of GetPdfInfo().
+  virtual void GetPdfInfo(
+      const std::vector<int32> &phones,
+      const std::vector<std::vector<std::pair<int32, int32> > > &pdf_class_pairs,
+      std::vector<std::vector<std::vector<std::pair<int32, int32> > > > *pdf_info)
       const;
 
  private:
   int32 N_;  //
   int32 P_;
   EventMap *to_pdf_;  // owned here.
 
+  // 'context' is the context-window of phones, of
+  // length N, with -1 for those positions where phones 
+  // that are currently unknown, treated as wildcards; at least 
+  // the central phone [position P] must be a real phone, i.e. 
+  // not -1. 
```

Comment:
I'm a little confused now about how you solved the issue about word-boundaries.
The comment says that -1 is used for positions where we have not yet expanded the phone.
However, the actual code seems to be using 0 for the "not-expanded"/wildcard phone.
But for the left and right context when we hit the edge of the file, the phone should be set to zero.  Not including zero as an option for positions other than position P_ would cause a problem.  But I don't see how this can happen, given that you are (in the code) using 0 for the wildcard.